### ISSUE TITLE
websockets' cookies/005.html doesn't pass in any browser

### ISSUE THREAD

**Issue by jdm** [w3c/web-platform-tests#2625]
There are things about the test that look suspicious, like the `onclose` handler of `ws` supposedly being unreachable, yet the test directly calls `ws.close()`. However, Firefox, Chrome, Safari, and Servo all have an error event dispatched before that, which suggests that the output from the websocket server is invalid for some reason.

cc @zcorpan

**Comment by jdm** [comment]
Yes, commenting out the `request.connection.write` line in `set-cookie_http_wsh.py` allows the test to reach the "unreachable" ws.onclose handler :)

**Comment by jdm** [comment]
The test passes if I replace the manual header creation in `set-cookie_http_wsh.py` with `request.extra_headers.append(('Set-Cookie', 'ws_test_'+(url_parts.query or '')+'=test; Path=/; HttpOnly'))`.

**Comment by jdm** [comment]
Looks like `simple_handshake_wsh.py` has a similar issue (meaning `opening-handshake/005.html` fails in all browsers too).

**Comment by zcorpan** [comment]
It passes in Opera 12 :-)

I believe `onclose` should indeed be unreachable *before the test `done()`*. Invoking `close()` doesn't ever synchronously fire `close` per spec.

I *think* you're supposed to get a `message` event before getting an `error` event, also (and so `onerror` should be unreachable, too), but I would be happy to review a pull request that cleans up the handlers. Testing the order of events should probably be separate tests.

**Comment by bobthekingofegypt** [comment]
Hello @zcorpan ,

I was looking into this test for servo when jdm pointed me to this ticket.  I identified a couple of points in https://github.com/servo/servo/issues/9772, including the one mentioned above about the handshake.

The problem with doing the connection write is that is doesn't override the normal handshake response which makes the first frame recieved on the socket the default HTTP upgrade response from pywebsocket, which doesn't contain valid framing for a websocket message.  Chrome complains with Opcode -1, servo throws a protocol error with an illegal control frame, firefox seems to just accept it (but does fail the test).  I've attached a screenshot from wireshark to try show this.  

Frame 652 is the custom handshake response
Frame 654 is the default handshake response, it's on the websocket channel and it's confused wireshark into thinking the H is the opcode for the response.

![closebreakdown](https://cloud.githubusercontent.com/assets/332141/14364911/f53fa1b0-fd02-11e5-8a6e-e45198d2b453.png)

The second problem is in the function 

```python
def web_socket_transfer_data(request):
    while True:
        return
```

The return on this functions first call will trigger pywebsocket to send a close message as soon as the handshake is complete.  This allows the browser to begin the close process including calling onclose.
I could be wrong, but it is my belief that this puts you into a non-deterministic state, depending on the execution of threads and incoming messages the close my be acknowledged before the incoming message on the echo socket is received.  In servo this creates an intermittent failure.

Leaving the test intact, what would you think of changing the handler to something like this

```python
def web_socket_do_extra_handshake(request):
    url_parts = urlparse.urlsplit(request.uri)
    request.extra_headers.append(('Set-Cookie', 'ws_test_'+(url_parts.query or '')+'=test; Path=/; HttpOnly'))

def web_socket_transfer_data(request):
    msgutil.receive_message(request)
```

Here we allow pywebsocket to handle the handshake for us, and we use the blocking receive_message to hold the socket open until the client closes it.

**Comment by zcorpan** [comment]
Changing the handler SGTM.

However, I think we should have a separate test that does these weird things with double handshake and immediate close, to get interoperable behavior. The spec should give deterministic behavior, I believe. The response is valid up to and including the first message, so the connection should open, you should get a message, and then the JS calls `close()`, and then it gets more garbage but that doesn't matter since you're in `CLOSING` state. If this doesn't work then there's a bug in the implementation.
